src/pint/utils.py

```diff
@@ -1595,3 +1596,38 @@ def print_color_examples():
                     end="",
                 )
             print("")
+
+
+class GroupIterator:
+    """An iterator to step over identical items in a :class:`numpy.ndarray`
+
+    Example
+    -------
+    This will step over all of the observatories in the TOAs.
+    For each iteration it gives the observatory name and the indices that correspond to it
+
+        t = pint.toa.get_TOAs("grouptest.tim")
+        I = GroupIterator(t["obs"])
+        for o, i in I:
+            print(f"{o} {i}")
+
+    """
+
+    def __init__(self, items):
+        self.items = items
+        self.unique_items = np.unique(items)
+        self.indices = []
+        for item in self.unique_items:
+            self.indices.append(np.where(self.items == item)[0])
```

Comment:
It might be preferable to use boolean indexing (just don't apply `np.where`) - among other things it enforces the length of the array you're indexing into. And you can always use `np.where` or `np.nonzero` later.

Reply:
I had done it like that earlier, but found that I consistently needed the extra `where` step.

Reply:
Boolean indexing should work on TOAs objects, both for read and write, I think?